src/composed/message/decrypt.rs

```diff
@@ -3,26 +3,73 @@ use log::debug;
 use crate::crypto::sym::SymmetricKeyAlgorithm;
 use crate::errors::Result;
 use crate::packet::SymKeyEncryptedSessionKey;
-use crate::types::{Mpi, SecretKeyRepr, SecretKeyTrait};
+use crate::types::{EskType, Mpi, SecretKeyRepr, SecretKeyTrait};
 
 /// Decrypts session key using secret key.
-pub fn decrypt_session_key<F, L>(locked_key: &L, key_pw: F, mpis: &[Mpi]) -> Result<PlainSessionKey>
+pub fn decrypt_session_key<F, L>(
+    locked_key: &L,
+    key_pw: F,
+    values: &EskBytes,
+    typ: EskType,
+) -> Result<PlainSessionKey>
 where
     F: FnOnce() -> String,
     L: SecretKeyTrait<Unlocked = SecretKeyRepr>,
 {
     debug!("decrypt session key");
 
-    locked_key.unlock(key_pw, |priv_key| {
-        let (key, sym_alg) = priv_key.decrypt(mpis, locked_key.fingerprint().as_bytes())?;
-        // TODO: handle other versions
-        Ok(PlainSessionKey::V4 { key, sym_alg })
-    })
+    locked_key.unlock(key_pw, |priv_key| priv_key.decrypt(values, typ, locked_key))
+}
+
+/// Values comprising the encrypted session key
+///
+/// FIXME: extend for algorithm specific values? (and/or v3 vs. v6)
+/// FIXME: naming?
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum EskBytes {
+    Rsa {
+        mpi: Mpi,
+    },
+    Elgamal {
+        first: Mpi,
+        second: Mpi,
+    },
+    Ecdh {
+        public_point: Mpi,
+        encrypted_session_key: Vec<u8>,
+    },
+    X25519 {
+        // Ephemeral X25519 public key (32 bytes).
+        ephemeral: [u8; 32],
+
+        // Encrypted and wrapped session key.
+        session_key: Vec<u8>,
+
+        // Set for v3 only (the sym_algo is not encrypted with the session key for X25519)
+        sym_alg: Option<SymmetricKeyAlgorithm>,
+    },
+    X448 {
+        // Ephemeral X448 public key (56 bytes).
+        ephemeral: [u8; 56],
+
+        // Encrypted and wrapped session key.
+        session_key: Vec<u8>,
+
+        // Set for v3 only (the sym_algo is not encrypted with the session key for X448)
+        sym_alg: Option<SymmetricKeyAlgorithm>,
+    },
+    Other,
 }
 
 /// Decrypted session key.
+///
+/// A v4 session key can be used with a SED, or a v1 SEIPD.
+/// A v6 session key can only be used with a v2 SEIPD.
+///
+/// https://www.rfc-editor.org/rfc/rfc9580.html#name-packet-versions-in-encrypte
 #[derive(derive_more::Debug, Clone, PartialEq, Eq)]
 pub enum PlainSessionKey {
+    /// Obtained from a v3 PKESK or a v4 SKESK
```

Comment:
is this correct `v3 PKESK or a v4 SKESK` ?

Reply:
I think so. For some historical reason, there is no v4 PKESK :shrug:

- PKESK was at v3 in RFC 2440/4880, and now went to v6.
- SKESK was at v4 in RFC 2440/4880, and now went to v6.

Reply:
In the newly defined type `EskType` I've called one variant `V3_4` (read: V3 "PKESK or V4 SKESK").
If we are happy with this (kind of weird) name, then we could also rename `PlainSessionKey::V4` to `V3_4` for consistency?

Whatever we do, these versions are just weird. I guess there is no super clean and neat way to name them.

Reply:
sounds all horrible, I guess `V3_4` helps a bit, to highlight you need to look at the docs

Reply:
changed in 401f356fc138c1572066e211389357f669effc35, and i added a bit of docs